sync/debug_mutex.go

```diff
@@ -0,0 +1,123 @@
+package xsync
+
+import (
+	"bytes"
+	"fmt"
+	"runtime"
+	"sync"
+	"time"
+)
+
+// DebugMutex is a RWMutex that tracks its ownership.
+type DebugMutex sync.RWMutex
+
+// RLock locks DebugMutex for reading.
+func (m *DebugMutex) RLock() { (*sync.RWMutex)(m).RLock() }
+
+// RUnlock undoes a single RLock call.
+func (m *DebugMutex) RUnlock() { (*sync.RWMutex)(m).RUnlock() }
+
+// Lock locks DebugMutex for writing.
+func (m *DebugMutex) Lock() {
+	(*sync.RWMutex)(m).Lock()
+	insert(m)
+}
+
+// Unlock unlocks DebugMutex for writing.
+func (m *DebugMutex) Unlock() {
+	remove(m)
+	(*sync.RWMutex)(m).Unlock()
+}
+
+// RLocker returns a Locker interface implemented via calls to RLock
+// and RUnlock.
+func (m *DebugMutex) RLocker() sync.Locker {
+	return (*rlocker)(m)
```

Comment:
Maybe just return `(*sync.RWMutex)(m).RLocker()` here so we don't need a separate `rlocker` type?